libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/internal/navigation/FreeDriveLocationUpdater.kt

```diff
@@ -0,0 +1,142 @@
+
+    fun start() {
+        if (future == null) {
+            locationEngine.requestLocationUpdates(locationEngineRequest, callback, null)
+            val handler = Handler(Looper.getMainLooper())
+            future = executorService.scheduleAtFixedRate({
+                if (rawLocation != null) {
+                    val enhancedLocation = getLocation(Date(), 0, rawLocation)
+                    handler.post {
+                        navigationEventDispatcher.onEnhancedLocationUpdate(
+                            enhancedLocation
+                        )
+                    }
+                }
+            }, 1500, 1000, TimeUnit.MILLISECONDS)
+        }
+    }
+
+    fun stop() {
+        future?.let {
+            stopLocationUpdates()
+        }
+    }
+
+    fun kill() {
+        future?.let {
+            stopLocationUpdates()
+            executorService.shutdown()
+        }
+    }
+
+    fun updateLocationEngine(locationEngine: LocationEngine) {
+        this.locationEngine = locationEngine
+        requestLocationUpdates()
+    }
+
+    fun updateLocationEngineRequest(request: LocationEngineRequest) {
+        this.locationEngineRequest = request
+        requestLocationUpdates()
+    }
+
+    private fun getLocation(date: Date, lagMillis: Long, rawLocation: Location?): Location {
+        val status = mapboxNavigator.retrieveStatus(date, lagMillis)
+        return getMapMatchedLocation(status, rawLocation)
+    }
+
+    private fun getMapMatchedLocation(
+        status: NavigationStatus,
+        fallbackLocation: Location?
+    ): Location {
+        val snappedLocation = Location(fallbackLocation)
+        snappedLocation.provider = "enhanced"
+        val fixLocation = status.location
+        val coordinate = fixLocation.coordinate
+        snappedLocation.latitude = coordinate.latitude()
+        snappedLocation.longitude = coordinate.longitude()
+        fixLocation.bearing?.let { snappedLocation.bearing = it }
+        snappedLocation.time = fixLocation.time.time
+        return snappedLocation
+    }
+
+    private fun stopLocationUpdates() {
+        locationEngine.removeLocationUpdates(callback)
+        future?.cancel(false)
+        future = null
+    }
+
+    private fun requestLocationUpdates() {
+        future?.let {
+            stop()
+            start()
+        }
+    }
+
+    private fun onLocationChanged(location: Location?) {
+        location?.let { currentLocation ->
+            rawLocation = currentLocation
+            mapboxNavigator.updateLocation(currentLocation)
+        }
+    }
+
+    private class CurrentLocationEngineCallback(locationUpdater: FreeDriveLocationUpdater) :
+        LocationEngineCallback<LocationEngineResult> {
+
+        private val updaterWeakReference: WeakReference<FreeDriveLocationUpdater> =
+            WeakReference(locationUpdater)
+
+        override fun onSuccess(result: LocationEngineResult) {
+            updaterWeakReference.get()?.let { locationUpdater ->
```

Comment:
NIT: `result.lastLocation` can be moved up and this doesn't have to be wrapped in `let`.